expandgrid_vs_CJ.Rmd

```diff
@@ -0,0 +1,65 @@
+---
+title: "Justification for data.table import"
+output: html_document
+date: "2023-06-30"
+---
+
+```{r Define both functions}
+library(data.table)
+library(microbenchmark)
+
+gh_fill_dt <- function(geohashes, precision) {
+  if (uniqueN(nchar(geohashes)) > 1) {
+    stop("Input Geohashes must all have the same precision level.")
+  }
+  if (sum(grepl("['ailo]", geohashes)) > 0) {
+    stop("Invalid Geohash; Valid characters: [0123456789bcdefghjkmnpqrstuvwxyz]")
+  }
+
+  new_levels <- precision - nchar(geohashes[1])
+
+  base32 <-
+    unlist(strsplit("0123456789bcdefghjkmnpqrstuvwxyz", split = ""))
+
+  grid <-
+    do.call(data.table::CJ, append(list(geohashes), replicate(new_levels, base32, FALSE)))
+
+  do.call(paste0, grid)
+
+}
+
+
+gh_fill_base <- function(geohashes, precision) {
+  if (uniqueN(nchar(geohashes)) > 1) {
+    stop("Input Geohashes must all have the same precision level.")
+  }
+  if (sum(grepl("['ailo]", geohashes)) > 0) {
+    stop("Invalid Geohash; Valid characters: [0123456789bcdefghjkmnpqrstuvwxyz]")
+  }
+
+  new_levels <- precision - nchar(geohashes[1])
+
+  base32 <-
+    unlist(strsplit("0123456789bcdefghjkmnpqrstuvwxyz", split = ""))
+
+  grid <-
+    do.call(expand.grid, append(list(geohashes), replicate(new_levels, base32, FALSE)))
+
+  do.call(paste0, grid)
+
+}
+```
+
+
+
+```{r create test vector}
+test_vector <- gh_fill_dt(c("9w", "9x"), 4L) # Length of 2048
+```
+
+```{r benchmark test}
+microbenchmark(gh_fill_dt(test_vector, 6L))
+microbenchmark(gh_fill_base(test_vector, 6L))
```

Comment:
On my machine the speedup is large but not massive (30%):

```
Unit: milliseconds
                          expr      min        lq      mean    median       uq
   gh_fill_dt(test_vector, 6L) 497.8866  723.1318  797.3614  793.1192  871.561
 gh_fill_base(test_vector, 6L) 814.1263 1031.9389 1225.3436 1160.6757 1457.280
      max neval
 1176.892   100
 1788.559   100
```

Anyway, data.table is not complex as far as dependencies go... I lean towards allowing it.